core/trino-main/src/main/java/io/trino/sql/ExpressionUtils.java

```diff
@@ -265,6 +276,40 @@ public static Expression filterConjuncts(Metadata metadata, Expression expressio
         };
     }
 
+    /**
+     * Returns whether expression is effectively literal. An effectitvely literal expression is a constant value, or null,
+     * in either {@link Literal} form, or other form returned by {@link LiteralEncoder}.
+     */
+    public static boolean isEffectivelyLiteral(PlannerContext plannerContext, Session session, Expression expression)
+    {
+        if (expression instanceof Literal) {
+            return true;
+        }
+        if (expression instanceof Cast) {
+            return ((Cast) expression).getExpression() instanceof Literal
+                    // a Cast(Literal(...)) can fail, so this requires verification
+                    && simplifiesToConstant(plannerContext, session, expression);
+        }
+        if (expression instanceof FunctionCall) {
+            QualifiedName functionName = ((FunctionCall) expression).getName();
+            if (isResolved(functionName)) {
+                ResolvedFunction resolvedFunction = plannerContext.getMetadata().decodeFunction(functionName);
+                return LITERAL_FUNCTION_NAME.equals(resolvedFunction.getSignature().getName());
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean simplifiesToConstant(PlannerContext plannerContext, Session session, Expression expression)
```

Comment:
Please rename this method to reflect its purpose better. IIUC, the purpose is ensuring that the expression does not fail.

Reply:
renamed to `constantExpressionEvaluatesSuccessfully`